text/0032-fog-compatible-gift-codes.md

```diff
@@ -0,0 +1,201 @@
+check of the temporary account, since the outputs were not indexed by Fog.
+
+This could be worked around, by making the person who creates the temporary account create it as a Fog account.
+However, there could be several deployments of Fog, and this basically means that the creator of the gift code needs to know
+what software the recipient is using, which is counter to the goal of the feature.
+
+In later versions of transfer payloads, the `TxOut` public key was included.
+
+```
+/// Message encoding a private key and a UTXO, for the purpose of
+/// giving someone access to an output. This would most likely be
+/// used for gift cards.
+message TransferPayload {
+    /// [Deprecated] The root entropy, allowing the recipient to spend the money.
+    /// This has been replaced by a BIP39 entropy.
+    bytes root_entropy = 1 [deprecated=true];
+
+    /// The public key of the UTXO to spend. This is an optimization, meaning
+    /// the recipient does not need to scan the entire ledger.
+    external.CompressedRistretto tx_out_public_key = 2;
+
+    /// Any additional text explaining the gift
+    string memo = 3;
+
+    /// BIP39 entropy, allowing the recipient to spend the money.
+    /// When deriving an AccountKey from this entropy, account_index is always 0.
+    bytes bip39_entropy = 4;
+}
+```
+
+It is possible to use the untrusted fog ledger endpoint to lookup `TxOut` by their public key,
+but this is not private, and the fog operator can see which `TxOut` you looked up. This puts the
+gift codes outside the normal threat-model for Fog transfers.
+
+We propose to fix this in a quite simple way:
+
+Instead of transfering private keys to an entire account, we just transfer ownership of a specific `TxOut`.
+We give them private keys and secrets connected to this `TxOut`, but which don't reveal anything about our account private keys.
+
+* To enable someone to spend a TxOut, we just have to give them the `onetime_private_key` of the `TxOut`.
+* To enable someone to see the unblinded amount of the TxOut, we just have to give them the `shared_secret` of the `TxOut`.
+* To enable someone to find the TxOut in the ledger (even a Fog user), we just have to tell them the global index in the blockchain.
+
+Given this data, a Fog user who receives the payload can ask Fog ledger for a merkle proof of membership for the `TxOut`.
+They can get the entire `TxOut` this way, match it against the shared secret and so unblind the amount commitment, and given the one-time
+private key, they can spend it as normal using the transaction builder.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A `TxOutGiftCode` is a payload which confers ownership of an individual `TxOut` in the blockchain,
+and is a little less than 80 bytes in size.
+
+* To enable someone to spend a TxOut, we just have to give them the `onetime_private_key` of the `TxOut`.
```

Comment:
As cited elsewhere, there's a race to spend. Notably a lot of clients tend to select txos based on optimization based algos. Clients would have to implement some kind txo locking mechanism with a ridiculously sized tombstone block (maybe like as in real gift cards - it has a redemption window) see: https://github.com/mobilecoinofficial/full-service/pull/270 in order to implement this in a way that's safe. 

There's a small enough amount of client languages I think where it *might* be reasonable for us to provide a set of traits/interfaces (java/android)/protocols (swift/ios) like "CanLockTxo" for clients that they'd need to implement to be able to use this feature. It could be enforced in some cases for things that use the specific APIs we provide. For Signal it would be feasible for us to PR these into their client to provide this functionality. 

Question is: is that work worth it?

Reply:
You'd have to share the information about the "reserved" TxOuts between the linked clients, while you could do it on Signal via the message channel, other clients though, unfortunately, can't do it. Hence we can't have a guarantee that any other client that has access to this account won't accidentally spend your gift code before it was redeemed.